### DISCUSSION TITLE
Self-installing Yocto using USB

### DISCUSSION THREAD

**Discussion by jsvalverde** [OE4T/meta-tegra#1481] [Q&A]
Hi everyone,

I'm interested in creating a Yocto Jetson image, which has to be able to "self-install" on a Jetson AGX Orin using a USB.
By self-installing I'm talking about a kind of installer program that could be booted from USB like an Ubuntu installer. The idea is to plug the USB, choose the boot device at the start menu, and start the flashing and booting process to load the Yocto-generated image.

I have seen several guides on the internet where something similar is done for x86 based systems, where they use the WIC format to generate the image containing both FAT boot partition and EXT4 image partition.
For example, I found this guide: https://www.thegoodpenguin.co.uk/blog/self-installing-yocto-image-from-a-usb-drive/, but it doesn't seem pretty compatible with how the Jetson works.

Reviewing several forums regarding the use of WIC on Jetson boards, I couldn't find any concrete cases where its use has been successful, and I was wondering what other alternatives may be useful to accomplish this task.

Since initrd flashing does a similar process when flashing to an external device in the sense that it first boots a minimal image and then it starts to flash the actual storage device with the new content I was wondering if this should be the path for accomplishing the feature we want.

Does anyone have some hints on how to achieve this?

Thanks in advance.

**Comment by dwalkes** [comment]
> Since initrd flashing does a similar process when flashing to an external device in the sense that it first boots a minimal image and then it starts to flash the actual storage device with the new content

Initrd flashing actually works a little different in that the device itself doesn't actually do all of the flashing, but rather exports some devices to the host for flashing.  See the sequence [here](https://github.com/OE4T/meta-tegra/wiki/initrd-flashing-support#flashing-after-build).  It's never been completely clear to me why it was done this way or whether there's a way to fix this so the host interaction is not necessary, but I believe what we do essentially matches they way NVIDIA does this and the only way they support initial device programming.

I think you are essentially looking for the "recovery image" concept discussed [here](https://github.com/OE4T/meta-tegra/wiki/initrd-flashing-support#possible-future-enhancements) which might be able to leverage some of the work for initrd flashing but will ultimately be a bit different.

I don't have any specific hints about how to do this other than to understand the internals of the [initrd-flash](https://github.com/OE4T/meta-tegra/blob/2ef932f823987bb67a65d969a344948e97332f60/recipes-bsp/tegra-binaries/tegra-helper-scripts/initrd-flash.sh#L29) and [init-flash](https://github.com/OE4T/meta-tegra/blob/2ef932f823987bb67a65d969a344948e97332f60/recipes-core/initrdscripts/tegra-flash-init/init-flash.sh#L50) scripts and how to decouple the host requirements for exporting devices to the host for flashing from these as an option.

**Comment by userian** [comment]
@jsvalverde - Did you make any progress on this? I'm working on a similar setup.

**Comment by dlheard17** [comment]
Also curious if you were able to do this @jsvalverde @userian

**Comment by userian** [comment]
I didn't get it working but I didn't spend much time on it as we ended up
utilizing PXE boot for flashing
